x-pack/plugins/security/server/usage_collector/security_usage_collector.test.ts

```diff
@@ -0,0 +1,425 @@
+      .getCollectorByType('security')
+      ?.fetch(clusterClient.asScoped().callAsCurrentUser);
+
+    expect(usage).toEqual({
+      auditLoggingEnabled: false,
+      accessAgreementEnabled: false,
+      authProviderCount: 0,
+      enabledAuthProviders: [],
+      loginSelectorEnabled: false,
+      httpAuthSchemes: [],
+    });
+  });
+
+  it('does not report disabled auth providers', async () => {
+    const config = createSecurityConfig(
+      ConfigSchema.validate({
+        authc: {
+          providers: {
+            basic: {
+              basic: {
+                order: 0,
+              },
+              disabledBasic: {
+                enabled: false,
+                order: 1,
+              },
+            },
+            saml: {
+              disabledSaml: {
+                enabled: false,
+                realm: 'foo',
+                order: 2,
+              },
+            },
+          },
+        },
+      })
+    );
+    const usageCollection = usageCollectionPluginMock.createSetupContract();
+    const license = createSecurityLicense({ isLicenseAvailable: true });
+    registerSecurityUsageCollector({ usageCollection, config, license });
+
+    const usage = await usageCollection
+      .getCollectorByType('security')
+      ?.fetch(clusterClient.asScoped().callAsCurrentUser);
+
+    expect(usage).toEqual({
+      auditLoggingEnabled: false,
+      accessAgreementEnabled: false,
+      authProviderCount: 1,
+      enabledAuthProviders: ['basic'],
+      loginSelectorEnabled: false,
+      httpAuthSchemes: ['apikey'],
+    });
+  });
+
+  it('reports the types and count of enabled auth providers', async () => {
+    const config = createSecurityConfig(
+      ConfigSchema.validate({
+        authc: {
+          providers: {
+            basic: {
+              basic: {
+                order: 0,
+                enabled: false,
+              },
+            },
+            saml: {
+              saml1: {
+                realm: 'foo',
+                order: 1,
+              },
+              saml2: {
+                realm: 'bar',
+                order: 2,
+              },
+            },
+            pki: {
+              pki1: {
+                enabled: true,
+                order: 3,
+              },
+            },
+          },
+        },
+      })
+    );
+    const usageCollection = usageCollectionPluginMock.createSetupContract();
+    const license = createSecurityLicense({ isLicenseAvailable: true });
+    registerSecurityUsageCollector({ usageCollection, config, license });
+
+    const usage = await usageCollection
+      .getCollectorByType('security')
+      ?.fetch(clusterClient.asScoped().callAsCurrentUser);
+
+    expect(usage).toEqual({
+      auditLoggingEnabled: false,
+      accessAgreementEnabled: false,
+      authProviderCount: 3,
+      enabledAuthProviders: ['saml', 'pki'],
+      loginSelectorEnabled: true,
+      httpAuthSchemes: ['apikey'],
+    });
```

Comment:
Nit: these test cases would be a bit easier to understand and maintain if test cases with a particular focus area were 1. grouped together in `describe` blocks, and 2. only checking the variables that should change, for example here:
```suggestion
    expect(usage).toEqual(
      expect.objectContaining({
        authProviderCount: 3,
        enabledAuthProviders: ['saml', 'pki'],
      })
    );
```

Reply:
I'm not opposed to grouping in `describe` blocks, but I rather liked asserting on everything in here. Our `xpack.security.*` config settings are rather complex, and involves custom logic outside of the traditional validation offered by `@kbn/config-schema`.

I viewed this as a way to assert that the entire flow from raw user config -> transformed config -> telemetry reporting was functioning correctly.

You could rightly argue that this is the responsibility of an integration test, but I'm not aware of a mechanism to do this with telemetry reporting today. If this starts to get out of hand then I'm happy to revisit, but I think it's tolerable for the time being, especially if I group the tests by focus area.

What do you think?

Reply:
Fair enough, you've convinced me!